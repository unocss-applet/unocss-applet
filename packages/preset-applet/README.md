# @unocss-applet/preset-applet

The Applet preset for [UnoCSS](https://github.com/unocss/unocss), fork from [@unocss/preset-uno](https://github.com/unocss/unocss/tree/main/packages/preset-uno) and modified to transform some CSS selector.

## Install

```bash
npm i @unocss-applet/preset-applet --save-dev # with npm
yarn add @unocss-applet/preset-applet -D # with yarn
pnpm add @unocss-applet/preset-applet -D # with pnpm
```

## Usage

```ts
import { defineConfig } from 'unocss'
import { presetApplet } from '@unocss-applet/preset-applet'

export default defineConfig({
  presets: [
    presetApplet(),
  ],
})
```

## Type Declarations

```ts
// PresetUnoOptions https://github.com/unocss/unocss/blob/main/packages/preset-uno/src/index.ts#L9
// PresetMiniOptions https://github.com/unocss/unocss/blob/main/packages/preset-mini/src/index.ts#L33-L73
export interface PresetAppletOptions extends PresetUnoOptions {
  /**
   * Unsupported characters in applet, will be added to the default value
   * @default ['.', ':', '%', '!', '#', '(', ')', '[', '/', ']', ',', '$', '{', '}', '@', '+', '^', '&', '<', '>', '\'', '\\', '"', '?']
   */
  unsupportedChars?: string[]

  /**
   * Space Between and Divide Width Elements
   * @default ['view', 'button', 'text', 'image']
   */
  betweenElements?: string[]
}
```

## Change

- The `*` selector will be replaced with `page` in the generated class name.
- the unsupported characters in applet will be replaced with `_a_` in the generated class name.

## Example

### Using in with `class`

#### without

```html
<div class="py-3.5 grid-cols-[0.7fr_repeat(7,1fr)]">
  py-3
</div>
```

#### with

```html
<div class="py-3_a_5 grid-cols-_a_0_a_7fr_repeat_a_7_a_1fr_a__a_">
  py-3
</div>

<style>
.grid-cols-_a_0_a_7fr_repeat_a_7_a_1fr_a__a_ {
  grid-template-columns: 0.7fr repeat(7,1fr);
}

.py-3_a_5 {
  padding-top:0.875rem;padding-bottom: 0.875rem;
}
</style>
```

### Using with string

If you want to ignore a string, add a prefix(default `applet-ignore:`), and the plugin will automatically ignore the string and delete the prefix.

#### without

```html
<script setup lang="ts">
  const bg = 'bg-[hsl(2.7,81.9%,69.6%)]'
  const bg2 = 'applet-ignore: bg-[hsl(2.7,81.9%,69.6%)]'
</script>
```

#### with

```html
<script setup lang="ts">
  const bg = 'bg-_a_hsl_a_2_a_7_a_81_a_9_a__a_69_a_6_a__a__a_'
  const bg2 = 'bg-[hsl(2.7,81.9%,69.6%)]'
</script>

<style>
.bg-_a_hsl_a_2_a_7_a_81_a_9_a__a_69_a_6_a__a__a_ {
  --un-bg-opacity: 1;
  background-color: hsla(2.7, 81.9%, 69.6%, var(--un-bg-opacity));
}
</style>
```

## License

MIT License &copy; 2022-PRESENT [Neil Lee](https://github.com/zguolee)
