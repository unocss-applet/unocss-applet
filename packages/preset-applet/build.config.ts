import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  entries: [
    'src/index',
  ],
  clean: true,
  declaration: true,
  externals: [
    'unocss',
    '@unocss/preset-uno',
    '@unocss/preset-mini',
    '@unocss/rule-utils',
  ],
})
