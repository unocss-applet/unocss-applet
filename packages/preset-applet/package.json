{
  "name": "@unocss-applet/preset-applet",
  "type": "module",
  "version": "0.4.13",
  "description": "The Applet preset for [UnoCSS](https://github.com/unocss/unocss).",
  "author": "Neil Lee <zguolee@foxmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/unocss-applet/unocss-applet/tree/main/packages/preset-applet#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/unocss-applet/unocss-applet.git",
    "directory": "packages/preset-applet"
  },
  "bugs": "https://github.com/unocss-applet/unocss-applet/issues",
  "keywords": [
    "unocss",
    "unocss-preset",
    "uni-app"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.cjs",
      "import": "./dist/index.mjs"
    },
    "./*": "./*"
  },
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub"
  },
  "dependencies": {
    "@unocss-applet/utils": "workspace:*",
    "@unocss/core": "^0.51.13",
    "@unocss/preset-mini": "^0.51.13",
    "@unocss/preset-uno": "^0.51.13"
  }
}
