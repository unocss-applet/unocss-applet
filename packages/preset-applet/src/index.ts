import type { Preset } from 'unocss'
import { presetUno } from 'unocss'
import type { PresetUnoOptions, Theme } from '@unocss/preset-uno'
import { normalizePreflights } from '@unocss/preset-mini'
import { encodeNonLatin } from '@unocss-applet/utils'
import { appletPreflights, defaultPreflights } from './preflights'

export type { Theme }

// PresetUnoOptions https://github.com/unocss/unocss/blob/main/packages/preset-uno/src/index.ts#L9
export interface PresetAppletOptions extends PresetUnoOptions {
  /**
   * Enable applet
   * @default true
   */
  enable?: boolean

  /**
   * Unsupported characters in applet, will be added to the default value
   * @default ['.', ':', '%', '!', '#', '(', ')', '[', '/', ']', ',', '$', '{', '}', '@', '+', '^', '&', '<', '>', '\'']
   */
  unsupportedChars?: string[]
}

export default function presetApplet(options: PresetAppletOptions = {}): Preset<Theme> {
  const enable = options.enable ?? true
  if (!enable)
    return presetUno(options)

  options.dark = options.dark ?? 'class'
  options.attributifyPseudo = options.attributifyPseudo ?? false
  options.preflight = options.preflight ?? true
  options.variablePrefix = options.variablePrefix ?? 'un-'

  const UNSUPPORTED_CHARS = ['.', ':', '%', '!', '#', '(', ')', '[', '/', ']', ',', '$', '{', '}', '@', '+', '^', '&', '<', '>', '\'']
  if (options.unsupportedChars)
    UNSUPPORTED_CHARS.push(...options.unsupportedChars)

  const ESCAPED_ESCAPED_UNSUPPORTED_CHARS = UNSUPPORTED_CHARS.map(char => `\\\\\\${char}`)
  const charTestReg = new RegExp(`${ESCAPED_ESCAPED_UNSUPPORTED_CHARS.join('|')}`)
  const charReplaceReg = new RegExp(`${ESCAPED_ESCAPED_UNSUPPORTED_CHARS.join('|')}`, 'g')

  function unoCSSToAppletProcess(str: string) {
    if (charTestReg.test(str))
      str = str.replace(charReplaceReg, '_a_')
    return str
  }

  return {
    ...presetUno({ ...options, preflight: false }),
    name: 'unocss-preset-applet',
    preflights: options.preflight
      ? normalizePreflights(enable ? appletPreflights : defaultPreflights, options.variablePrefix)
      : [],
    postprocess: [
      (util) => {
        if (enable && util.selector) {
          util.selector = unoCSSToAppletProcess(util.selector)
          util.selector = encodeNonLatin(util.selector)
        }
        return util
      },
    ],
  }
}
