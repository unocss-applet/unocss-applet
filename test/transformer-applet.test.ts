import type { UnocssPluginContext } from '@unocss/core'
import { presetApplet, transformerApplet } from '@unocss-applet/preset-applet'
import { createGenerator } from '@unocss/core'
import MagicString from 'magic-string'
import { describe, expect, it } from 'vitest'

const transformer = transformerApplet()

describe('transformer-applet', async () => {
  const uno = await createGenerator({
    presets: [
      presetApplet({
        presetOptions: {
          dark: 'media',
        },
        unsupportedChars: ['~', ' '],
      }),
    ],
  })

  async function transform(code: string) {
    const s = new MagicString(code)
    await transformer.transform(s, '', {
      tokens: new Set<string>(),
      uno,
    } as UnocssPluginContext)
    return s.toString()
  }

  it('basic', async () => {
    const transformTargets = [
      '-ml-1.5 ml-1.5 -mt-2',
      'bg-[url(https://api.iconify.design/carbon:bat.svg?color=red)]',
    ]

    const result = await transform(transformTargets.join(' '))
    expect(result).toMatchInlineSnapshot(`"-ml-1_a_5 ml-1_a_5 -mt-2 bg-_a_url_a_https_a__a__a_api_a_iconify_a_design_a_carbon_a_bat_a_svg_a_color_a_red_a__a_"`)
  })
})
