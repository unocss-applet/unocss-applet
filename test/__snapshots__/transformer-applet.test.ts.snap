// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformer-applet > basic 1`] = `
{
  "code": "<!-- eslint-disable @typescript-eslint/quotes -->
<script setup lang=\\"ts\\">
import { ref } from 'vue'

const bg = 'bg-_a_hsl_a_2_a_7_a_81_a_9_a__a_69_a_6_a__a__a_'

const icon = \\"i-carbon:campsite\\"
const bgIgnore = 'applet-ignore: bg-_a_hsl_a_2_a_7_a_81_a_9_a__a_69_a_6_a__a__a_'
const index = 1
const customClass = 'text-red'
const bool = ref<boolean>()
const border = ref(true)
const fixed = ref(true)
</script>

<template>
  <div class=\\"text-center aaa bg-_a__a_var_a_\\" p=\\"4\\">
    <div text=\\"4xl\\" class=\\"rotate-180 i-carbon-campsite\\" :class=\\"bg\\" />
    <div :class=\\"[border && 'u-border-top', fixed && 'u-tabbar--fixed']\\" />
    <div class=\\"border bg-blue-200 font-(light mono) \\" :cc=\\"customClass\\">
      <div class=\\"hover:(!bg-gray-600 text-red font-bold)\\" text=\\"#fff\\">
        {{ 'applet-ignore: hover:(!bg-gray-600 text-red font-bold)' }}
      </div>
    </div>
    <div :class=\\"\`p-2_a_5 \${bool ? 'p-0_a_5' : ''}\`\\" m-2 :hover-class=\\"['_a_bg-green', bool ? '' : 'text-red']\\">
      class=\\"hover_a_bg-green\\"
    </div>
    <div flex=\\"~ col gap-1\\" class=\\"p-1\\" items-center :class=\\"bool ? 'text-yellow-500 px-2_a_5' : ''\\">
      <div :class=\\"icon\\" inline-block color=\\"blue\\" text=\\"xl !red\\" />
      <div bg=\\"green-(!200 800)\\">
        {{ \`index\${index + 1}\` }}{{ \`index\` }}
      </div>
    </div>
    <div flex=\\"~ col\\" un-b=\\"~ green\\" :class=\\"\`bg-color-\${bg}\`\\">
      <div text-right h-10 flex=\\"1\\" text=\\"red\\" :class=\\"{ 'text-sm': index > 0 }\\">
        0123456789
      </div>
      <div
        h-10 flex=\\"1\\" :class=\\"[index > 1 ? 'text' : '']\\"
        text=\\"blue dark:(red !bold)\\" :style=\\"[index > 1 ? '' : '']\\"
        :type=\\"index > 1\\"
      >
        {{ bgIgnore }}
      </div>
    </div>
    <div class=\\"bg-_a_url_a_https_a__a__a_static_a_runoob_a_com_a_images_a_demo_a_demo2_a_jpg_a__a_\\" w-40 h-20 ma color=\\"white\\" bg=\\"center cover\\">
      {{ 'applet-ignore: bg-_a_url_a_https_a__a__a_static_a_runoob_a_com_a_images_a_demo_a_demo2_a_jpg_a__a_' }}
    </div>
    <div class=\\"m-0_a_5 p-1 text-2xl\\" :class=\\"bool ? '' : 'text-yellow-500 p-2_a_5'\\">
      abckefghijklmnopqrstuvwxyz
    </div>
    <div block class=\\"size-10\\" size=\\"20\\" />
    <div class=\\"[&_*]:after:right-4 _a__a__a_div_a__a_bg-red [&+*]:text-xl\\" />
  </div>
</template>
",
  "css": "/* layer: default */
.fixed{position:fixed;}
._a__a__\\\\*_a__a_after_a_right-432*::after{right:1rem;}
.m-0_a_5{margin:0.125rem;}
.m-2,
[m-2=\\"\\"]{margin:0.5rem;}
.ma,
[ma=\\"\\"]{margin:auto;}
.block,
[block=\\"\\"]{display:block;}
.inline-block,
[inline-block=\\"\\"]{display:inline-block;}
.h-10,
[h-10=\\"\\"]{height:2.5rem;}
.h-20,
[h-20=\\"\\"]{height:5rem;}
.w-40,
[w-40=\\"\\"]{width:10rem;}
[flex~=\\"\\\\~\\"]{display:flex;}
[flex~=\\"\\\\3132\\"]{flex:1 1 0%;}
[flex~=\\"col\\"]{flex-direction:column;}
.rotate-180{--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-rotate:180deg;transform:translateX(var(--un-translate-x)) translateY(var(--un-translate-y)) translateZ(var(--un-translate-z)) rotate(var(--un-rotate)) rotateX(var(--un-rotate-x)) rotateY(var(--un-rotate-y)) rotateZ(var(--un-rotate-z)) skewX(var(--un-skew-x)) skewY(var(--un-skew-y)) scaleX(var(--un-scale-x)) scaleY(var(--un-scale-y)) scaleZ(var(--un-scale-z));}
.items-center,
[items-center=\\"\\"]{align-items:center;}
.gap-1,
[flex~=\\"gap-1\\"]{grid-gap:0.25rem;gap:0.25rem;}
.border,
[un-b~=\\"\\\\~\\"]{border-width:1px;}
[un-b~=\\"green\\"]{--un-border-opacity:1;border-color:rgba(74,222,128,var(--un-border-opacity));}
._a__a__a_div_a__a_bg-red>div{--un-bg-opacity:1;background-color:rgba(248,113,113,var(--un-bg-opacity));}
._a_bg-green{--un-bg-opacity:1 !important;background-color:rgba(74,222,128,var(--un-bg-opacity)) !important;}
.bg-_a__a_var_a_{background-color:var(--\\\\{var\\\\});}
.bg-_a_hsl_a_2_a_7_a_81_a_9_a__a_69_a_6_a__a__a_{--un-bg-opacity:1;background-color:hsla(2.7,81.9%,69.6%,var(--un-bg-opacity));}
.bg-blue-200{--un-bg-opacity:1;background-color:rgba(191,219,254,var(--un-bg-opacity));}
.hover_a_bg-green:hover{--un-bg-opacity:1;background-color:rgba(74,222,128,var(--un-bg-opacity));}
.bg-_a_url_a_https_a__a__a_static_a_runoob_a_com_a_images_a_demo_a_demo2_a_jpg_a__a_{--un-url:url(https://static.runoob.com/images/demo/demo2.jpg);background-image:var(--un-url);}
[bg~=\\"cover\\"]{background-size:cover;}
[bg~=\\"center\\"]{background-position:center;}
.p-0_a_5{padding:0.125rem;}
.p-1{padding:0.25rem;}
.p-2_a_5{padding:0.625rem;}
[p~=\\"\\\\3432\\"]{padding:1rem;}
.px-2_a_5{padding-left:0.625rem;padding-right:0.625rem;}
.text-center{text-align:center;}
.text-right,
[text-right=\\"\\"]{text-align:right;}
._a__a__a_\\\\*_a__a_text-xl+*,
[text~=\\"xl\\"]{font-size:1.25rem;line-height:1.75rem;}
.text-2xl{font-size:1.5rem;line-height:2rem;}
.text-sm{font-size:0.875rem;line-height:1.25rem;}
[text~=\\"\\\\3432xl\\"]{font-size:2.25rem;line-height:2.5rem;}
[color~=\\"blue\\"],
[text~=\\"blue\\"]{--un-text-opacity:1;color:rgba(96,165,250,var(--un-text-opacity));}
[color~=\\"white\\"],
[text~=\\"_a_fff\\"]{--un-text-opacity:1;color:rgba(255,255,255,var(--un-text-opacity));}
.text-red,
[text~=\\"red\\"]{--un-text-opacity:1;color:rgba(248,113,113,var(--un-text-opacity));}
.text-yellow-500{--un-text-opacity:1;color:rgba(234,179,8,var(--un-text-opacity));}
[text~=\\"_a_red\\"]{--un-text-opacity:1 !important;color:rgba(248,113,113,var(--un-text-opacity)) !important;}
.size-10{width:2.5rem;height:2.5rem;}
[size~=\\"\\\\32320\\"]{width:5rem;height:5rem;}",
}
`;
